### PR TITLE
Switch the atomic implementation to use Clang  builtins.

### PR DESCRIPTION
**Pull request by AlexVlx**
Atomics need not be grabbed from HC, but can rather rely on builtins, which we lower correctly nowadays.

### CODE REVIEW

File: include/hip/hcc_detail/hip_atomic.h [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+    while (tmp < val) { tmp = atomicCAS(address, tmp, val); }
+
+    return tmp;
+}
+
+__device__
+inline
+unsigned int atomicInc(unsigned int* address, unsigned int val)
+{
+    extern unsigned int __builtin_amdgcn_atomic_inc(
```
Comment (yxsamliu) [inline include/hip/hcc_detail/hip_atomic.h]: hip-clang needs __device__ here

File: include/hip/hcc_detail/hip_atomic.h [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+
+    return __builtin_amdgcn_atomic_inc(
+        address, val, __ATOMIC_RELAXED, 1 /* Device scope */, false);
+}
+
+__device__
+inline
+unsigned int atomicDec(unsigned int* address, unsigned int val)
+{
+    extern unsigned int __builtin_amdgcn_atomic_dec(
```
Comment (yxsamliu) [inline include/hip/hcc_detail/hip_atomic.h]: need __device__

**Comment by yxsamliu** [comment]
LGTM other than the two comments. I tested that this makes HIP-Example reduction pass with hip-clang.

Great work! Thanks!

**Comment by AlexVlx** [comment]
@yxsamliu whoops, apologies, forgot about them pesky `__device__`s; added, thank you!

File: include/hip/hcc_detail/hip_atomic.h [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+unsigned long long atomicAdd(
+    unsigned long long* address, unsigned long long val)
+{
+    return __atomic_fetch_add(address, val, __ATOMIC_RELAXED);
+}
+__device__
+inline
+float atomicAdd(float* address, float val)
+{
+    unsigned int* uaddr{reinterpret_cast<unsigned int*>(uaddr)};
```
Comment (yxsamliu) [inline include/hip/hcc_detail/hip_atomic.h]: I got warning when compiling hipCaffe:
 warning: variable 'uaddr' is uninitialized when used within its own initialization [-Wuninitialized]
    unsigned int* uaddr{reinterpret_cast<unsigned int*>(uaddr)};
                  ~~~~~                                 ^~~~~
It seems this line tries to initialize uaddr with its own value. Is this intended?

File: include/hip/hcc_detail/hip_atomic.h [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+        old = atomicCAS(uaddr, r, __float_as_uint(val + __uint_as_float(r)));
+    } while (r != old);
+
+    return __uint_as_float(r);
+}
+__device__
+inline
+double atomicAdd(double* address, double val)
+{
+    unsigned long long* uaddr{reinterpret_cast<unsigned long long*>(uaddr)};
```
Comment (yxsamliu) [inline include/hip/hcc_detail/hip_atomic.h]: same warning as above

**Comment by AlexVlx** [comment]
@yxsamliu not intended at all, those are pretty terrible typos on my part, thank you for spotting them; we shall have to revisit the unit test in the future since those shouldn't have gone through.

**Pull request merged**